### ISSUE TITLE
[SB2 13.5"] i915 freezes on resume from hibernate/suspend

### ISSUE THREAD

**Issue by leonm1** [jakeday/linux-surface#305]
Upon resume, my Surface Book 2 13.5" becomes extremely unstable, manifesting in random freezes for ~10 seconds each until a final freeze that requires a reboot to fix. The log only shows messages like the following, which seem to correspond with the freezes:

```
Nov 26 11:44:28 matthsurface kernel: i915 0000:00:02.0: Resetting rcs0 for hang on rcs0
```

I have the latest patches (including qzed's surfacegen5_acpi_notify), firmware, and script in /lib/systemd/system-sleep/sleep (to which I added i915 while troubleshooting, no effect).

The system is rock solid stable before hibernating or sleeping. Does anyone have any insight on something I might've forgotten?

Other details:
Distro: archlinux
uname -r: 4.18.8-1-surface4
Method of kernel install: this [PKGBUILD](https://gist.github.com/leonm1/8cf99d041d48567207360f44992e30f6) with qzed's patches

**Comment by qzed** [comment]
I assume this also happens without the battery patches/module?

**Comment by qzed** [comment]
Ok, can confirm this for Jake's 4.18.12 (without the battery patches) and my modified 4.18.19. A bit more details:

```
[  130.857920] [drm] GPU HANG: ecode 9:0:0xfffffffe, in Xwayland [1756], reason: hang on rcs0, action: reset
[  130.858195] i915 0000:00:02.0: Resetting rcs0 for hang on rcs0
[  130.869879] i915 0000:00:02.0: Resetting chip for hang on rcs0
[  130.883996] [drm] GuC: Loaded firmware i915/kbl_guc_ver9_39.bin (version 9.39)
[  130.884061] i915 0000:00:02.0: GuC firmware version 9.39
[  130.884062] i915 0000:00:02.0: GuC submission enabled
[  130.884064] i915 0000:00:02.0: HuC disabled
```

For me, this only occurs once on a few suspends and if it does, the system usually is responsive after a few more seconds, only sometimes breaking completely. There seem to be a few bug reports mentioning `hang on rcs0` on the kernel bugzilla, e.g. [this one](https://bugs.freedesktop.org/show_bug.cgi?id=104545) so that might be worth investigating.

**Comment by leonm1** [comment]
> For me, this only occurs once on a few suspends and if it does, the system usually is responsive after a few more seconds, only sometimes breaking completely. There seem to be a few bug reports mentioning `hang on rcs0` on the kernel bugzilla, e.g. [this one](https://bugs.freedesktop.org/show_bug.cgi?id=104545) so that might be worth investigating.

Looking at the actual bugzilla, I tried the firmware they recommend earlier. Jake's repo already contains the v1.04 firmware for the gpu, so it seems like a dead end for the surface.

I also tried getting the patches working on 4.20-rc4; however, there were many changes to the IPTS files and getting those working with the static contexts is proving a challenge. I noticed there might be an effort to make those patches work on 4.19? If there is a unified issue on that, I'd be happy to jump on and provide my efforts.

**Comment by qzed** [comment]
Thanks for your help!

> however, there were many changes to the IPTS files

Yeah, I had a quick look at 4.19 a while back, but the changes in the i915 stuff are quite significant (#303 should be the relevant issue for that). Once I noticed that for porting this would take me more than a day I gave up (for now), still have the battery module to finish first. Help for IPTS on 4.19 (and 4.20+) would be greatly appreciated!
